src/development/cplusplus-codeguide.rst

```diff
@@ -0,0 +1,296 @@
+ability to generate buildfiles for different development environments. 
+This allows developers the freedom to use whatever IDE they
+would like (e.g. Eclipse and Xcode) from the same CMake files.
+
+Unfortunately there is even less consensus in the community as to the most
+effective way to write CMake files, thought there are a lot of opinions. We
+have found a good distillation of some effective ideas can be found `here
+<https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1>`_. And GitLab
+has an introduction to `Modern CMake
+<https://cliutils.gitlab.io/modern-cmake/>`_.
+
+In general however we recommend Modern CMake (minimum version of 3.0). A lot of legacy
+projects will have used 2.8 - but we cannot recommend that new projects use
+this. It is no longer maintained and does not contain many features of modern
+(v3+ CMake). As far as we are aware CMake v3.0 is backward compatible
+with 2.8, so if you are onboarding code that uses 2.8 we strongly suggest you upgrade. 
+
+.. todo:: Maybe add some more advice? use targets not global settings etc ...
+
+Including the version number
+""""""""""""""""""""""""""""
+There is a template configuration file that demonstrates how to pull in the
+version number into the code base. We do this in a few  steps. First the Semantic
+Version number is held in plain text in version.txt. This is done so that it is
+easy to "bump" the version number on code changes. The second stage is the the
+CMakeLists.txt file reads this file and creates an internal version number
+using its contents. Finally the configuration file is used to generate a header
+file containing the version number that can be included by the code.
+
+Including External Projects
+"""""""""""""""""""""""""""
+We have including the GoogleTest framework as an external project, instead of a
+dependency to demonstrate one way to use external projects. 
+
+.. warning:: We have used a git submodule to do this. There may be a formal recomendation that git subtrees be used instead.
+
+Runtime Dependencies
+""""""""""""""""""""
+
+.. todo:: Add a simple description of dependency and version listing
+
+We consider that there are two principles of dependency management we would dictate:
+
+1) Dependency installation should be unambigigous and repeatable. This requires dependencies to be listed, together with a minimum, or required version number. 
+
+2) Dependencies should be project specific, there should be no reason for developers to pollute their environment just to build your package.
+
+We do not advocate a solution here - there is not a dominant package management system for C++. Indeed you could just pull and build all your deopendencies via external projects as the our GoogleTest example. Another option would simply be to list the dependencies in a text file *dependencies.txt* and allow the developer to use their preferred pacakge management system to install them.
+
```

Comment:
I think we need to be a lot more explicit and mandatory here.
An example dependencies file in the top level showing the required format with the CMakeLists.txt using that file to include the relevant cmake modules for each dependency. i.e. one source of infomation.

Reply:
as above - we need a c++ dependency feature to deal with this explicitly